crates/telio-pmtu/src/linux.rs

```diff
@@ -0,0 +1,378 @@
+        (libc::SOL_IP, libc::IP_RECVERR)
+    };
+    let val: c_int = 1;
+
+    if unsafe {
+        libc::setsockopt(
+            sock.as_raw_fd(),
+            lvl,
+            name,
+            &val as *const _ as *const _,
+            std::mem::size_of_val(&val) as _,
+        )
+    } == -1
+    {
+        return Err(io::Error::last_os_error());
+    }
+
+    Ok(())
+}
+
+fn read_pmtu(sock: &impl AsRawFd, is_ipv6: bool) -> io::Result<u32> {
+    let mut mtu: c_int = 0;
+    let mut size: libc::socklen_t = std::mem::size_of_val(&mtu) as _;
+
+    let (lvl, name) = if is_ipv6 {
+        (libc::IPPROTO_IPV6, libc::IPV6_MTU)
+    } else {
+        (libc::IPPROTO_IP, libc::IP_MTU)
+    };
+
+    unsafe {
+        if libc::getsockopt(
+            sock.as_raw_fd(),
+            lvl,
+            name,
+            &mut mtu as *mut _ as *mut _,
+            &mut size as *mut _,
+        ) == -1
+        {
+            Err(io::Error::last_os_error())
+        } else {
+            Ok(mtu as _)
+        }
+    }
+}
+
+fn read_error_pmtu(sock: &impl AsRawFd, is_ipv6: bool) -> io::Result<u32> {
+    let cmsg_lvl = if is_ipv6 {
+        libc::SOL_IPV6
+    } else {
+        libc::SOL_IP
+    };
+
+    unsafe {
+        let mut iovecbuf = [0u8; 8];
+
+        let mut iovec = libc::iovec {
+            iov_base: iovecbuf.as_mut_ptr().cast(),
+            iov_len: iovecbuf.len() as _,
+        };
+
+        let mut controlbuf = [0u8; 128];
+
+        let mut msghdr = libc::msghdr {
+            msg_name: std::ptr::null_mut(),
+            msg_namelen: 0,
+            msg_iov: &mut iovec,
+            msg_iovlen: 1,
+            msg_control: controlbuf.as_mut_ptr().cast(),
+            msg_controllen: controlbuf.len() as _,
+            msg_flags: 0,
+        };
+
+        loop {
+            if libc::recvmsg(
+                sock.as_raw_fd(),
+                &mut msghdr,
+                libc::MSG_DONTWAIT | libc::MSG_ERRQUEUE,
+            ) == -1
+            {
+                return Err(io::Error::last_os_error());
+            } else {
+                let mut cmsg = libc::CMSG_FIRSTHDR(&msghdr);
+
+                while !cmsg.is_null() {
+                    if (*cmsg).cmsg_level == cmsg_lvl && (*cmsg).cmsg_type == libc::IP_RECVERR {
+                        let err = *(libc::CMSG_DATA(cmsg) as *const libc::sock_extended_err);
+
+                        telio_log_debug!(
+                            "Extended sock error:\n  ee_errno   = {}\n  ee_origin  = {}\n  ee_type  = {}\n  ee_code  = {}\n  ee_pad   = {}\n  ee_info  = {}\n  ee_data  = {}",
+                            err.ee_errno,
+                            err.ee_origin,
+                            err.ee_type,
+                            err.ee_code,
+                            err.ee_pad,
+                            err.ee_info,
+                            err.ee_data,
+                        );
+
+                        if err.ee_errno == libc::EMSGSIZE as u32 {
```

Comment:
I might be missing some part, but I wonder whether it would make sense to take the last error of type EMSGSIZE in this case instead of first one? could it be possible, that some errors would be queued from either previous runs, or from this run, but different MTU attempt making the error stale?

Reply:
I don't think it is possible. We do the queue check every time we get an `recv()` error. Of course the kernel might decide whatever it wants but since we treat this as a hint rather the true PMTU then I don't see any problems with that.